### ISSUE TITLE
Add support for TDBU Blinds

### ISSUE THREAD

**Issue by jeffwelsch** [gormanb/homebridge-connector-hub#10]
**Describe Your Problem:**

I am attempting to configure connector-hub to manage 3 Day Blinds motorized roman shades.  I have an AMP Bridge (PN 62104000) and six shades today, with a seventh coming soon.  My shades include four top/bottom up/down and two bottom up/down only shades.  I am able to manage the two bottom up/down only shades only and the four top/bottom shades experience the failed to sanitize device state error.

The different shades also have different remote controls, if that matters.  The bottom up/down only shades use AMP-1SC-Remote and the top/bottom use AMP-ISC-Dual-Remote.  I notice in the logs the RSSI for these shades is not good, but they are installed next to shades that do work, which also have poor RSSI (based on the iOS App).

I did notice that the shades that function were listed as Roman Shades--hex digits and the ones that are not working are listed as TDBU Blinds 01-a4e57cbe69fb; where 01 seems to represent the shade number.

**Logs:**

[6/19/2023, 3:35:05 PM] [homebridge-connector-hub] Failed to sanitize device state: {
  msgType: 'ReadDeviceAck',
  mac: 'a4e57cbe69fb0001',
  deviceType: '10000001',
  msgID: '20230619153501280',
  data: {
    type: 9,
    exist_subid: 1,
    operation_B: 2,
    currentPosition_B: 0,
    currentState_B: 3,
    voltageMode: 1,
    batteryLevel_B: 831,
    wirelessMode: 1,
    RSSI: -98
  }
}
[6/19/2023, 3:35:05 PM] [homebridge-connector-hub] Failed to sanitize device state: {
  msgType: 'ReadDeviceAck',
  mac: 'a4e57cbe69fb0002',
  deviceType: '10000001',
  msgID: '20230619153501296',
  data: {
    type: 9,
    exist_subid: 1,
    operation_T: 2,
    operation_B: 2,
    currentPosition_T: 0,
    currentPosition_B: 0,
    currentState_T: 3,
    currentState_B: 3,
    voltageMode: 1,
    batteryLevel_T: 830,
    batteryLevel_B: 828,
    wirelessMode: 1,
    RSSI: -99
  }
}
[6/19/2023, 3:35:05 PM] [homebridge-connector-hub] Failed to sanitize device state: {
  msgType: 'ReadDeviceAck',
  mac: 'a4e57cbe69fb0004',
  deviceType: '10000001',
  msgID: '20230619153501330',
  data: {
    type: 9,
    exist_subid: 1,
    operation_B: 2,
    currentPosition_B: 0,
    currentState_B: 3,
    voltageMode: 1,
    batteryLevel_B: 833,
    wirelessMode: 1,
    RSSI: -98
  }
}
[6/19/2023, 3:35:05 PM] [homebridge-connector-hub] Failed to sanitize device state: {
  msgType: 'ReadDeviceAck',
  mac: 'a4e57cbe69fb0006',
  deviceType: '10000001',
  msgID: '20230619153501363',
  data: {
    type: 9,
    exist_subid: 1,
    operation_B: 2,
    currentPosition_B: 0,
    currentState_B: 3,
    voltageMode: 1,
    batteryLevel_B: 828,
    wirelessMode: 1,
    RSSI: -86
  }
}

**Plugin Config:**

        {
            "hubIps": [
                "192.168.1.127"
            ],
            "connectorKey": "real key is configured",
            "enableDebugLog": true,
            "platform": "ConnectorHub",
            "_bridge": {
                "username": "0E:D9:29:9E:EB:BE",
                "port": 30909
            }
        }

**Environment:**

* **Plugin Version**: homebridge-connector-hub v1.0.9
* **Homebridge Version**: v1.6.1
* **Node.js Version**: v16.15.0
* **NPM Version**: 8.5.5
* **Operating System**: Docker/Ubuntu

**Comment by jeffwelsch** [comment]
I've since updated to node v18.16.0 and npm 9.5.1 and the issue persists.

**Comment by gormanb** [comment]
Hi @jeffwelsch: the problem here is that the plugin doesn't account for the `_T` and `_B` suffixes on the properties returned from the hub to distinguish between the top and bottom blinds. This should hopefully be straightforward to fix - would you be willing to test out a couple of beta versions to make sure that they work as expected?

**Comment by jeffwelsch** [comment]
Thank you,

Yes, I would be happy to help.  I won't be able to test until earlier next week but will do anything I can do to help get this working -- it's working great for the other shades and I'm looking forward to complete coverage!

**Comment by gormanb** [comment]
Hi @jeffwelsch - I've just published version `1.1.0-beta.0`, which adds support for TDBU blinds. Your devices should now appear in Homekit as a separate accessory for each top down and bottom up blind. Please give it a try and let me know what happens!

**Comment by gormanb** [comment]
Quick followup @jeffwelsch: I spotted a small bug which could affect the operation of TDBU blinds, so I've published `1.1.0-beta.1` to correct it. Please test that version instead!

**Comment by jeffwelsch** [comment]
Thanks for the update.  I installed the beta package and it does seem to be working now.  This is now in my logs:

```
[7/6/2023, 1:24:27 PM] [homebridge-connector-hub] Updating position: [ 'TDBU Blinds Bottom-Up 01-a4e57cbe69fb', 97 ]
[7/6/2023, 1:24:27 PM] [homebridge-connector-hub] Updating battery: [ 'TDBU Blinds Bottom-Up 01-a4e57cbe69fb', 50 ]
[7/6/2023, 1:24:27 PM] [homebridge-connector-hub] Updating position: [ 'TDBU Blinds Bottom-Up 02-a4e57cbe69fb', 73 ]
[7/6/2023, 1:24:27 PM] [homebridge-connector-hub] Updating battery: [ 'TDBU Blinds Bottom-Up 02-a4e57cbe69fb', 100 ]
[7/6/2023, 1:24:27 PM] [homebridge-connector-hub] Updating position: [ 'Roman Blinds 03-a4e57cbe69fb', 100 ]
[7/6/2023, 1:24:27 PM] [homebridge-connector-hub] Updating battery: [ 'Roman Blinds 03-a4e57cbe69fb', 100 ]
[7/6/2023, 1:24:27 PM] [homebridge-connector-hub] Updating position: [ 'TDBU Blinds Bottom-Up 04-a4e57cbe69fb', 100 ]
[7/6/2023, 1:24:27 PM] [homebridge-connector-hub] Updating battery: [ 'TDBU Blinds Bottom-Up 04-a4e57cbe69fb', 100 ]
[7/6/2023, 1:24:27 PM] [homebridge-connector-hub] Updating position: [ 'Roman Blinds 05-a4e57cbe69fb', 100 ]
[7/6/2023, 1:24:27 PM] [homebridge-connector-hub] Updating battery: [ 'Roman Blinds 05-a4e57cbe69fb', 100 ]
[7/6/2023, 1:24:27 PM] [homebridge-connector-hub] Updating position: [ 'TDBU Blinds Bottom-Up 06-a4e57cbe69fb', 100 ]
[7/6/2023, 1:24:27 PM] [homebridge-connector-hub] Updating battery: [ 'TDBU Blinds Bottom-Up 06-a4e57cbe69fb', 100 ]
[7/6/2023, 1:24:27 PM] [homebridge-connector-hub] Updating position: [ 'TDBU Blinds Bottom-Up 07-a4e57cbe69fb', 85 ]
[7/6/2023, 1:24:27 PM] [homebridge-connector-hub] Updating battery: [ 'TDBU Blinds Bottom-Up 07-a4e57cbe69fb', 100 ]
[7/6/2023, 1:24:32 PM] [homebridge-connector-hub] Updating battery: [ 'Roman Blinds 03-a4e57cbe69fb', 100 ]
```

I won't be at that home until this weekend to fully test on the blinds.  I will report back when I am able to test.

**Comment by jeffwelsch** [comment]
I've been able to test more and it seems to be mostly working.  Two items:

First, I noticed this in my logs:

> homebridge.log:[7/8/2023, 8:44:00 PM] [homebridge-connector-hub] Periodic refresh failed: TDBU Blinds Bottom-Up 07-a4e57cbe69fb

I have that entry for every shade, but I also didn't have the child bridge connected, which I just remedied.  I enabled debug logging and will see if I get more detail should this happen again.

The second is a usability question.  These are top down/bottom up, but I cannot seem to find out how to lower the top.  I can open and close the shades, and can close them a specific percentage, but in the Home app I cannot open the top when the shade is closed.  The idea is to have the bottom 50% closed.  Am I missing a configuration to allow control of the top of the shade?

**Comment by gormanb** [comment]
Hi @jeffwelsch - thanks for reporting back! Regarding the `Periodic refresh failed` message - this isn't cause for concern, it just indicates that the blinds were slow to respond to one of the plugin's status update requests. These happen every 5 seconds, so the blinds' status will simply be updated on the next refresh. It only indicates a problem if you get the message **_every_** 5 seconds consistently, which would suggest that the hub is not contactable.

Regarding the fact that the top-down blinds are not accessible: I noticed in the log output you posted at the outset of this ticket that some messages contained information on both `currentPosition_T` and `currentPosition_B`, while others only mentioned `currentPosition_B`. I surmised that some of your blinds might have both bottom-up and top-down functions while others might be bottom-up only. But is it the case that ALL your blinds are both bottom-up and top-down?

If so, then it seems that the information from the hub might sometimes report only one of these, and sometimes both. Could you please turn on debug logging, open and close the bottom-up part of your blind, and post the resulting logs here?

**Comment by jeffwelsch** [comment]
Hi @gormanb 

I will ignore the periodic refresh update.  The hub certainly has flaky WiFi and RF connectivity so it would make sense it periodically has issues, thank you.  I do have both top down only and top down/bottom up shades and those without the currentPosition_T would be the bottom up only shades.

The experience I am missing is illustrated by the Hunter Douglas TPBU shade integration with homekit.  In the home application there are two sliders, one for the top down and one for the bottom up.  In the attached screenshot, 'Blinds' represents the bottom down portion and 'Blinds 2' represents the top up portion.  In this configuration the bottom blind is closed completely, and the top is 45% open, resulting in a privacy setting with the top open and the bottom closed.

![Screenshot 2023-07-10 at 5 10 12 PM](https://github.com/gormanb/homebridge-connector-hub/assets/6233633/3545327b-cfe6-4116-b40a-8723ff75d1c3)

**Comment by gormanb** [comment]
Hi @jeffwelsch - I've just published `1.1.0-beta.4` which should hopefully expose both top-down and bottom-up blinds. Each of these will be represented as a separate accessory. Please give it a try and let me know how it works - if the functionality is OK, perhaps we can investigate merging both TD and BU into a single accessory per your screenshot above.

**Comment by jeffwelsch** [comment]
OK, I have been able to get everything working properly, insofar as the connectivity to the hub will allow.  I am able to manipulate the top and bottom independently.  One request would be a way to reverse the control for the top down portion because the behavior of the top portion is the opposite of the bottom portion.  Specifically:

- Botton physically lowered, Home reports the shade is closed.
- Bottom raised, Home reports the shade is open.
- Top lowered, Home reports the shade is closed.
- Top raised, Home reports the shade is open.

Thanks for the help!

**Comment by gormanb** [comment]
Hi @jeffwelsch - thanks for the update, sounds like we're almost there 😄 I've just published `1.1.0-beta.5`, which should hopefully fix the issue you're having with the top-down components. Please give it a try and let me know how it goes!

**Comment by jeffwelsch** [comment]
@gormanb works great!  Now when the top is open, Homekit reports correctly. One nit, and I'm not sure you can do anything about this: the icon in Home reports the status correctly but the slider reflects the bottom shade.  I.e. when open, the top shade opens down and not up like the bottom shade does.  Can the slider control in Home also be reversed?

**Comment by gormanb** [comment]
@jeffwelsch: unfortunately, it looks like there's no way to invert the slider direction in the Home app. Glad to hear that everything is working aside from that; I've just published a full release of `1.1.0`, so please upgrade to that 😄

**Comment by gormanb** [comment]
Fixed as of commit 8cbb096.
